firedrake/cofunction.py

```diff
@@ -225,8 +225,12 @@ def assign(self, expr, subset=None, expr_from_assemble=False):
             return self.assign(
                 assembled_expr, subset=subset,
                 expr_from_assemble=True)
-
-        raise ValueError('Cannot assign %s' % expr)
+        elif expr == 0:
```

Comment:
We do the same for `ufl.classes.Zero` on line 199 so this should probably be combined. In general I am a bit concerned about this code because we do not tape `Assigner`. @Ig-dolci knows more I believe.

Reply:
I see. I copied this from `Function.assign`, and just noticed that `Function.assign` is decorated with `@FunctionMixin._ad_annotate_assign`. This is required for `bc.set(Cofunction, float)` if we want to set diagonal entries to 1.